jax/numpy/lax_numpy.py

```diff
@@ -2140,7 +2141,8 @@ def arange(start, stop=None, step=None, dtype=None):
   lax._check_user_dtype_supported(dtype, "arange")
   if stop is None and step is None:
     dtype = dtype or _dtype(start)
-    return lax.iota(dtype, start)  # avoids materializing
+    size = start if type(start) is Poly else lax.convert_element_type(start, onp.uint64)
```

Comment:
The latter case would make `size` a `DeviceArray`, which seems like something we don't want. (We'd want to use `onp` to control types here.) @JuliusKunze but why do we need this line at all?

Reply:
@mattjj According to the [XLA spec](https://www.tensorflow.org/xla/operation_semantics#iota) `lax.iota` is not supposed to accept `float` size, so I added conversion here and made `iota` more strict. 

In order to minimize the diff, I reverted the change for now and put the conversion to `int` to [where it was before](https://github.com/google/jax/pull/2800/files#diff-08adae3ce53df47932e06971edb68642R1237) (otherwise tests fail since arange accepts float). (`arange`/`iota` currently doesn't work with `shapecheck` anyway. That will be part of the next PR, so no need to check for `Poly` here for now.)